docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -110,6 +111,42 @@ We need to be able to process transactions and roll-back state updates if a tran
 
 We identified use-cases, where modules will need to save an object commitment without storing an object itself. Sometimes clients are receiving complex objects, and they have no way to prove a correctness of that object without knowing the storage layout. For those use cases it would be easier to commit to the object without storing it directly.
 
+
+### Remove MultiStore
+
+IAVL based store adds additional layer in the SDK store construction - the `MultiStore` structure. The multistore exists to support the modularity of the Cosmos SDK - each module is using it's own instance of IAVL, but in the current implementation, all instances share the same database.
+The latter indicates, however, that the implementation doesn't provide true modularity. Instead it causes problems related to race condition and sync problems (eg: [\#6370](https://github.com/cosmos/cosmos-sdk/issues/6370)).
+
+We propose to remove the MultiStore from the SDK, and use a single instance of `SC`. To improve usability, we should extend the `KVStore` interface with _prefix store_:
+
+```
+type KVStore interface {
+    ... current KVStore
+
+    WithPrefix(prefix string) KVStore
+}
+```
+
+The `WithPrefix` method will create a proxy object for the parent object and will prepend `prefix` to a key parameter of all key-value operations (similar to the `store/prefix.Store` implementation).
+The following invariant should be applied
+
+```
+for each OP in [Get Has, Set, ...]
+    store.WithPrefix(prefix).OP(key) == store.OP(prefix + key)
+```
+
+#### Optimization: use compression for prefix keys
+
+Moreover we can consider a compression of prefix keys using Huffman Coding. It will require a knowledge of used prefixes. And for best results it will need an frequency information for each prefix (how often objects are stored in the store under the same prefix key). With Huffman Coding the above invariant should have the following shape:
+
+```
+for each OP in [Get Has, Set, ...]
+    store.WithPrefix(prefix).OP(key) == store.OP(store.Code(prefix) + key)
+```
+
+Where `store.Code(prefix)` is a Huffman Code of `prefix` in the given `store`.
+
```

Comment:
@aaronc originally suggested to map a module "verbose" key to a 2 byte sequence prefix for key prefix compression, eg:
```
bank -> 0x0001
staking -> 0x0002
....
```
In both mechanisms we will need to assure that the created map will be stable (we will always construct the same mapping pairs).

NOTE: Both Huffman Codes and module key map compress the keys only for the `SS`. It's not needed for `SC` because in `SC` we always operate on a hash of a key.

Reply:
I think the Aaron is easier to manage. The limitation is that we bound the number of modules to 65536 (2^16) -  which is big enough to worry about now.

Reply:
To make this prefix model work effectively, I think there should be some stateful prefix map - possibly stored under some special restricted schema prefix (probably `0`) on the `RootStore`. We can use varint encoding and not require a restriction to 2 bytes, but also 2 bytes is probably fine.

In this restricted schema key-space, we would have a mapping from `key name` -> `compressed prefix`, which is itself prefixed to allow for other schema use cases.

How does that sound?

Reply:
How important is it to have this optimization at all?

Reply:
String prefixes could potentially be rather long (maybe full proto msg names eventually as discussed in #9156). I think it's relatively important.

Reply:
If we want to enable variable prefix length, then huffman coding is better, because it takes frequency into account and it is also protects against common prefixes.